File: src/nishikihebi/console.py
from typing import TextIO

from nishikihebi.env import load_env_var

RESET = "0"
BOLD = "1"
DIM = "2"
RED = "31"
GREEN = "32"
YELLOW = "33"
BLUE = "34"
MAGENTA = "35"
CYAN = "36"

_SECTION_WIDTH = 72


def color_enabled(stream: TextIO) -> bool:
    if load_env_var("NO_COLOR"):
        return False

    match (load_env_var("NISHIKIHEBI_COLOR") or "").strip().lower():
        case "never":
            return False
        case "always":
            return True

    isatty = getattr(stream, "isatty", None)
    return bool(isatty and isatty())


def style(text: str, *codes: str, stream: TextIO) -> str:
    if not codes or not color_enabled(stream):
        return text
    return f"\x1b[{';'.join(codes)}m{text}\x1b[{RESET}m"


def section(title: str, *, stream: TextIO) -> None:
    if len(title) >= _SECTION_WIDTH:
        heading = f"{title} "
    else:
        rule = "─" * (_SECTION_WIDTH - len(title) - 1)
        heading = f"{title} {rule}"
    print(file=stream)
    print(style(heading, BOLD, CYAN, stream=stream), file=stream)

File: packages/shared/console.ts
import { loadEnvVar } from "./env.js";

export interface OutputStream {
  write(chunk: string): void;
  readonly isTTY?: boolean | undefined;
}

export const RESET = "0";
export const BOLD = "1";
export const DIM = "2";
export const RED = "31";
export const GREEN = "32";
export const YELLOW = "33";
export const BLUE = "34";
export const MAGENTA = "35";
export const CYAN = "36";

const SECTION_WIDTH = 72;

export function colorEnabled(stream: OutputStream): boolean {
  if (loadEnvVar("NO_COLOR")) {
    return false;
  }
  switch ((loadEnvVar("ANGEL_COLOR") ?? "").trim().toLowerCase()) {
    case "never":
      return false;
    case "always":
      return true;
    default:
      return stream.isTTY === true;
  }
}

export function style(
  text: string,
  codes: readonly string[],
  stream: OutputStream,
): string {
  if (codes.length === 0 || !colorEnabled(stream)) {
    return text;
  }
  return `\x1b[${codes.join(";")}m${text}\x1b[${RESET}m`;
}

export function section(title: string, stream: OutputStream): void {
  const heading =
    title.length >= SECTION_WIDTH
      ? `${title} `
      : `${title} ${"─".repeat(SECTION_WIDTH - title.length - 1)}`;
  stream.write("\n");
  stream.write(`${style(heading, [BOLD, CYAN], stream)}\n`);
}
